### ISSUE TITLE
System.InvalidCastException: 'Specified cast is not valid.'

### ISSUE THREAD

**Issue by IeuanWalker** [matt-goldman/Maui.Plugins.PageResolver#28]
**Describe the bug**
Getting a cast exception when trying to pass parameters - 
```
[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidCastException: Specified cast is not valid.
[mono-rt]    at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetServices(IServiceProvider provider, Type serviceType)
[mono-rt]    at Maui.Plugins.PageResolver.NavigationExtensions.IsRegisteredDependency(IServiceProvider serviceProvider, Type type)
[mono-rt]    at Maui.Plugins.PageResolver.NavigationExtensions.<>c__DisplayClass7_0.<ParametersMatchConstructors>b__0(ParameterInfo p)
[mono-rt]    at System.Linq.Enumerable.WhereArrayIterator`1[[System.Reflection.ParameterInfo, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
[mono-rt]    at System.Collections.Generic.LargeArrayBuilder`1[[System.Reflection.ParameterInfo, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].AddRange(IEnumerable`1 items)
[mono-rt]    at System.Collections.Generic.EnumerableHelpers.ToArray[ParameterInfo](IEnumerable`1 source)
[mono-rt]    at System.Linq.Enumerable.ToArray[ParameterInfo](IEnumerable`1 source)
[mono-rt]    at Maui.Plugins.PageResolver.NavigationExtensions.ParametersMatchConstructors(Type type, Object[] parameters)
[mono-rt]    at Maui.Plugins.PageResolver.NavigationExtensions.CreatePageWithViewModel[DisplayPage](IServiceProvider serviceProvider, Type viewModelType, Object[] parameters)
[mono-rt]    at Maui.Plugins.PageResolver.NavigationExtensions.ResolvePage[DisplayPage](Object[] parameters)
[mono-rt]    at Maui.Plugins.PageResolver.NavigationExtensions.<PushAsync>d__2`1[[App.Routes.WasteCalendar.Pages.DisplayPage, App, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
[mono-rt]    at App.Routes.WasteCalendar.Pages.AddressSearchPage.BtnNext_Clicked(Object sender, EventArgs e) in D:\App\Maui\FCMPOCO\App\Routes\WasteCalendar\Pages\AddressSearchPage.xaml.cs:line 24
[mono-rt]    at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
[mono-rt]    at Android.App.SyncContext.<>c__DisplayClass2_0.<Post>b__0() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.App/SyncContext.cs:line 36
[mono-rt]    at Java.Lang.Thread.RunnableImplementor.Run() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Thread.cs:line 36
[mono-rt]    at Java.Lang.IRunnableInvoker.n_Run(IntPtr jnienv, IntPtr native__this) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net8.0/android-34/mcw/Java.Lang.IRunnable.cs:line 84
[mono-rt]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V(_JniMarshal_PP_V callback, IntPtr jnienv, IntPtr klazz) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 22
```

**To Reproduce**
Navigating to a new page I am doing this - 
```csharp
await Navigation.PushAsync<DisplayPage>(_viewModel.SelectedAddress.Uprn, _viewModel.SelectedAddress.Address);
```

The new page has 2 injectable classes - 
![image](https://github.com/matt-goldman/Maui.Plugins.PageResolver/assets/6544051/ebaecb95-37cf-4266-accf-99cbd0ae6c47)

And the ViewModel for that page, has 1 injectable class + the 2 parameters being passed through -
![image](https://github.com/matt-goldman/Maui.Plugins.PageResolver/assets/6544051/5315a6eb-64d2-497c-b815-81f3ae046f7f)

All are being registered and the mapping is done manually - 
![image](https://github.com/matt-goldman/Maui.Plugins.PageResolver/assets/6544051/0c8dfdf9-c30d-4008-b4ac-51ca2befdd3c)

**Additional context**
Removing the 2 parameters (`double uprn, string address`), I don't get the exception.

**Comment by IeuanWalker** [comment]
@matt-goldman this line in the `NavigationExtensions` class is throwing the exception when the type is `Double`

![image](https://github.com/matt-goldman/Maui.Plugins.PageResolver/assets/6544051/9b9503ce-32e0-49b8-bff0-90ad243a2a2d)

**Comment by matt-goldman** [comment]
Hey @IeuanWalker thanks for this! Absolutely very happy for you to submit a PR, but I think I would prefer a slightly different approach. We can filter out primitives and some other obvious exclusions first to avoid the exception altogether, and then probably catch `Exception` more broadly.

```csharp
private static bool IsRegisteredDependency(IServiceProvider serviceProvider, Type type)
{
    if (type.IsPrimitive || type == typeof(string) || type.IsAbstract || type.IsValueType)
    {
        return false;
    }

    try
    {
        var services = serviceProvider.GetServices(type);
        return services.Any();
    }
    catch (Exception)
    {
        return false;
    }
}
```

Also based on #27 I'm planning to add an `[Ignore]` attribute (I've actually encountered a scenario where I needed it myself). So when I add that I'll also filter ignored types as well, and will move `if (type.IsPrimitive || type == typeof(string) || type.IsAbstract || type.IsValueType)` to another `IsIgnoredType` method.

Anyway for now what do you think of the above approach? Can you check it against your code?

**Comment by IeuanWalker** [comment]
@matt-goldman ye sounds good to me. Will get a PR sorted :)

**Comment by IeuanWalker** [comment]
@matt-goldman Had to remove the `|| type.IsAbstract` it comes back as true for interfaces, so would fail to find DI services that use an interface

**Comment by matt-goldman** [comment]
Fixed by @IeuanWalker in #29
